tensorflow_federated/python/core/impl/executors/federating_executor.py

```diff
@@ -486,6 +489,131 @@ def validate_executor_placements(cls, executor_placements):
                   'Unsupported cardinality for placement "{}": {}.'.format(
                       pl, pl_cardinality))
 
+  async def _trusted_aggregator_generate_keys(self):
+    
+    @computations.tf_computation()
+    def generate_keys():
+      pk, sk = easy_box.gen_keypair()
+      return pk.raw, sk.raw
+
+    fn_type = generate_keys.type_signature
+    fn = generate_keys._computation_proto
+
+    aggregator_ex = self._get_child_executors(placement_literals.AGGREGATOR,
+                                              index=0)
+
+    result = await aggregator_ex.create_call(
+        await aggregator_ex.create_value(fn, fn_type)
+    )
+
+    pk = await aggregator_ex.create_selection(result, 0)
+    sk = await aggregator_ex.create_selection(result, 1)
+
+    pk_fed_val = await self._place(pk, placement_literals.AGGREGATOR)
+    pk_fed_val = await self.federated_broadcast(pk_fed_val)
+    sk_fed_val = await self._place(sk, placement_literals.AGGREGATOR)
+
+    return pk_fed_val, sk_fed_val
+
+  async def _zip_val_key(self, val, key):
+
+    # zip values and keys EagerValues on each client using 
+    # their respective eager executor.
+    val_key_zips = []
+    key_placement = key.type_signature.placement.name
+    for i in range(len(val)):
+      client_ex = self._get_child_executors(placement_literals.CLIENTS, index=i)
+      val_eager_val = val[i]
+      if key_placement == 'CLIENTS':
+        key_eager_val = key.internal_representation[i]
+      elif key_placement == 'AGGREGATOR':
+        key_eager_val = key.internal_representation[0]
+      val_key_zip = await client_ex.create_tuple(
+          anonymous_tuple.AnonymousTuple([(None, val_eager_val), 
+                                          (None, key_eager_val)])
+      )
+
+      val_key_zips.append(val_key_zip)
+
+    return val_key_zips
+
+  async def _encrypt_client_tensors(self, arg, pk_a):
+
+    input_tensor_type = arg.type_signature[0].member
+    pk_a_tensor_type = pk_a.type_signature.member
+
+    @computations.tf_computation(input_tensor_type, 
+                                 pk_a_tensor_type)
+    def encrypt_tensor(plaintext, pk_a):
+
+      pk_a = easy_box.PublicKey(pk_a)
+      pk_c, sk_c = easy_box.gen_keypair()
+      nonce = easy_box.gen_nonce()
+      ciphertext, mac = easy_box.seal_detached(plaintext, 
+                                               nonce, 
+                                               pk_a, 
+                                               sk_c
+        )
+
+      return ciphertext.raw, mac.raw, pk_c.raw, nonce.raw
+
+    fn_type = encrypt_tensor.type_signature
+    fn = encrypt_tensor._computation_proto
+    val_type = arg.type_signature[0]
+    val = arg.internal_representation[0]
+
+    val_key_zipped = await self._zip_val_key(val, pk_a)
```

Comment:
could `_zip` be used here instead? in what way is `_zip_val_key` different?

Reply:
I don't have a good reason, except that when I originally worked on it I couldn't get code snippet working with `federated_zip` to figure out what was `_zip` expected as input (`arg`) and also validate it was the expected behaviour. Basically what was the arg type_signature and internal_representation. I could check again on Monday.

Reply:
maybe have a second look yes. if it makes sense to get rid of `_zip_val_key` then that'd be great; and also fine if we can't as long as we understand why

Reply:
Ok I am using now `_zip` to zip keys and values. I still have `_zip_val_key` because zip and val needs to be `NamedTuple` wrapped in `FederatedValue`. I had to modify a bit the code in `federated_reduce` when zipping the encrypted values and the secret aggregator key. This `_zip` happens on the `AGGREGATORS`. In `_zip` there is a cardinality check. If the op is not placed on the client, it expects only one item. In our case if we have 3 clients, we end up with 3 encrypted values on the `AGGREGATORS` and one secret key. So I had to loop through the encrypted values to zip them with the key.